### ISSUE TITLE
GRPC writing large data ignores isReady flag and eventually fails

### ISSUE THREAD

**Issue by BryanCutler** [OpenLiberty/open-liberty#23544]
When reading from a stream, then writing the message back to the server I have noticed a couple issues. I am using the Apache Arrow Flight client to read/write streams of data. After a while, the write stream starts to slow down to a crawl before stopping altogether. At that point the server sockets start to timeout. Before writing a message, the Arrow Flight client checks the grpc Stream `isReady()` to make sure the transport can accept the next message. This call always returns true when connecting to our Liberty server with the grpc feature enabled, even just before the stream completely stops. Because of that, there is no backpressure happening and the client does not throttle data being written. I have also tried this without Liberty and it works as expected, `isReady()` will return false and cause the client to slow down much earlier in the stream.

I have also noticed that even when the client cancels the call, the data already sent will continue to flow (very slowly) for even as long as 10 minutes. The server should receive a callback that the client has cancelled and can then stop the incoming stream, but this never happens. I confirmed that without Liberty it does get the callback.

**Steps to Reproduce**  
I can not share server code, but I could provide client code with the Apache Arrow Flight client.

**Expected behavior**  
Backpressure should be respected when writing, and the callback for client cancelling the stream should be sent.

**Diagnostic information:**  
 - OpenLiberty Version: [e.g. 21.0.0.8 - 21.0.0.10]
 - Affected feature(s) [e.g. mpHealth-3.0]
 - Java Version: [i.e. full output of `java -version`]
 - server.xml configuration (WITHOUT sensitive information like passwords)
 - If it would be useful, upload the messages.log file found in `$WLP_OUTPUT_DIR/messages.log`

**Additional context**

**Comment by BryanCutler** [comment]
I am wondering if there is any server configuration that I could set to control the buffering/backpressure during writes? cc @hutchig who helped with a prior issue.

**Comment by hutchig** [comment]
So I am investigating 3 possibilities currently:
1 - the gRPC CVE maintenance I have done has all been to do with [improving performance](https://github.com/protocolbuffers/protobuf/commit/a3888f53317a8018e7a439bac4abeb8f3425d5e9) of large data flows - I imagine that Bryan will use the 'latest' gRPC when he rolls his own so this is a prompt to bring the library used to the latest.
2 - It may be that the servlet based gRPC layer we have somehow does not pass back the back pressure as well as the 'standard' netty based gRPC
3 - It may just be a plain old omission/error in our code.
I am familiarizing myself with the gRPC codepaths and investigating, and plan to address 1 regardless now.

**Comment by BryanCutler** [comment]
Thanks for taking a look @hutchig , let me know if there is something I can do to help narrow the possibilities down.

**Comment by BryanCutler** [comment]
@hutchig following up on this as it is still an issue for us. Has there been any investigation or plans to implement any of the possible fixes above?

**Comment by hutchig** [comment]
Hi Bryan, apologies for slow update. I did have a look and the intent is to replace the code we have at https://github.com/OpenLiberty/open-liberty/tree/integration/dev/io.openliberty.grpc.1.0.internal/src/io/grpc/servlet with the equivalent from google with an upgrade to the recent https://github.com/grpc/grpc-java/releases/tag/v1.53.0 which has 
“servlet: Implement gRPC server as a Servlet ([#8596](https://github.com/grpc/grpc-java/pull/8596))”
https://github.com/grpc/grpc-java/pull/8596 which is the upstream equivalent of the code that we have in Liberty (which
is the crux of why gRPC in Liberty is 'different' from vanilla gRPC.) We will do this in #19217 but keep this one alive until
your problem is resolved.

**Comment by BryanCutler** [comment]
Thanks, I'll follow the issue above and see if it resolves the problem

**Comment by niloc132** [comment]
Hi, I'm the author of the linked grpc-servlet PR (though note that I am _not_ the author of most of the code, I just got it tidied up and ready for merge). I had need to check something on the PR itself, and so noticed the link made to this issue. We also ended up with a fork of the upstream PR, and we have several other fixes, including around an isReady check when writing metadata. 

The code in https://github.com/OpenLiberty/open-liberty/blob/integration/dev/io.openliberty.grpc.1.0.internal/src/io/grpc/servlet/ suggests that this is a much earlier version of the PR - there were definitely some fixes made to waiting until it is safe to do an async write.

From your notes above, I understand that you're using this for arrow flight. As it happens, we're using it for the same purpose, though are not using the default FlightBindingService from flight-core, but have implemented our own. We are currently still using our fork of grpc-servlet (available in maven central as `io.deephaven:deephaven-grpc-servlet-jakarta:0.26.1`), but intend to upstream at least some of the enhancements we've made.

Please feel free to reach out on github or directly if you have any questions, or find your issue isn't resolved by the upstream work.

**Comment by hutchig** [comment]
Thank you Colin @niloc132 . I am glad to hear others are working with the grpc-servlet. Although OpenLiberty is completely free Open Source and gets 100% of any updates from the IBM commercial WebSphere Liberty development (where most of my work, in other components, is) - it is embedded in a commercial product so I will be very very careful about making use of any Upstream/Fork in order to completely comply with any associated [license.](https://github.com/deephaven/deephaven-core/blob/3c4bea471d1aa66415bcd61c9eb21a1b8ee52caf/grpc-java/grpc-servlet-jakarta/LICENSE) I would encourage any fixes to become PRs upstream to grpc-java. We are very much encouraged to do that in IBM but completely understand it is a subtle topic that is different for each  org :-)

**Comment by niloc132** [comment]
I note that the grpc-java files already in this project appear to be copied from the original PR I based my work on, which marked those files as apache licensed - and in this repo they remain apache licensed (see https://github.com/OpenLiberty/open-liberty/blob/dbed0ecbdaf8c2ef62e1fcda25c3b4cd161ab609/dev/io.openliberty.grpc.1.0.internal/src/io/grpc/servlet/GrpcServlet.java#L2-L14, etc). In the same way, our fork of those files is also apache licensed (as you noted), as are other related files that we hope to likewise contribute, including a grpc-web proxy (intended to make it possible to use grpc and Arrow's Flight possible from a browser). So I believe we are already fully compatible?

**Comment by hutchig** [comment]
Yes I think we are, I just wanted to record that I would check the LICENSE carefully first and abide by it. Most of OpenLiberty is http://www.eclipse.org/legal/epl-2.0/ but those files retain the original license as they are slight modifications to fit with our runtime environment of the PR as it was at that time. If the grpc-servlet PR had been merged when we did the work originally we would have probably 'patched' the grpc classes as part of building our OSGi bundle of gRPC-java but it was just a source PR then.
